proposals/0000-lexical-scope.md

```diff
@@ -0,0 +1,586 @@
+
+The JEP also requires that unbound values error at evaluation time, not
+at compile time.  This enables implementations to bind an initial (global)
+scope when a query is evaluated.  This is something that other query languages
+provide, and is useful to define fixed queries that only vary by the value
+of specific variables.  We'll look at a few examples.
+
+First, consider a command line utility, let's called it `jp`, that accepts a
+path to a file containing a JMESPath query and reads an input JSON document
+through stdin.  This command line utility could offer a `--params` option that
+allows a user to pass in an initial scope.  For example:
+
+*myquery.jmespath*
+
+```
+results[*].[name, uuid, $hostname]
+```
+
+A user could then use this CLI to retrieve JSON data and filter it:
+
+```
+$ curl https://myapi/info/$HOSTNAME | \
+    jp --filename myquery.jmespath --params '{"hostname": "$HOSTNAME"}'
+```
+
+In this case the JMESPath expression does not need to change and can be
+shared with other people, and still include data that's specific to your
+machine.
+
+Another example would be where JMESPath is used in some shared definition file.
+Suppose we had a file that defines how to make an API request, and specifies a
+condition we'd like to meet based on the output response.  We want to describe
+that the expected output depends on the input provided.  This is how we can
+describe this:
+
+```json
+{"GroupActive": {
+   "operation": "DescribeGroups",
+   "acceptors": {
+     "argument": "Response[].[length(Instances[?State=='Active']) == length($params.GroupNames)",
+     "matcher": "path"
+   }
+}}
+```
+
+This is saying that we should invoke the `DescribeGroups` operations with a
+list of group names, and that we want to check that the response contains a
+list of `Instances` with `State == 'Active'` whose length matches the length of
+the params group names.  You could now bind the user provided params as the
+initial scope of `{"params": inputParams}` and code generate something like
+this (using the python JMESPath library in this example):
+
+```python
+def wait(user_params):
+  response = client.DescribeGroups(user_params)
+  expected = jmespath.compile(
+    "Response[].[length(Instances[?State=='Active']) "
+    "== length($params.GroupNames)"
+  )
+  result = expected.search(
+    response,
+    # This is the new part, give queries access to the user params
+    # via the $params variable.
+    scope={'params': user_params},
+)
+  if result:
+    return "SomeSuccessResponse"
+  return "SomeFailureResponse"
+
+# User can invoke this via:
+wait({"GroupNames": ["group1", "group2", "group3"]})
+```
+
+This JEP does not require that implementation provide this capability of
+passing in an initial scope, but by requiring that undefined variable
+references are runtime errors it enables implementations to provide this
+capability.  Implementations are also free to provide an opt-in "strict"
```

Comment:
Cool, makes sense. I'd probably add some kind of predefined-params option to the linters in https://github.com/awslabs/smithy/tree/main/smithy-jmespath/src/main/java/software/amazon/smithy/jmespath.